test/DataCollectors/TraceDataCollector.UnitTests/VangurdTests.cs

```diff
@@ -0,0 +1,189 @@
+            this.vangurdCommandBuilderMock = new Mock<IVangurdCommandBuilder>();
+            this.collectorUtilityMock  = new Mock<ICollectorUtility>();
+
+            this.vanguard = new Vanguard(this.collectorUtilityMock.Object, this.vangurdCommandBuilderMock.Object, this.processJobObject);
+            var guid = Guid.NewGuid();
+            this.sessionName = VangurdTests.SessionNamePrefix + guid;
+            this.configFileName = string.Format(VangurdTests.ConfigFileNameFormat, Path.GetTempPath(), guid);
+            this.outputDir = Path.GetDirectoryName(this.configFileName);
+            Directory.CreateDirectory(outputDir);
+            this.outputFileName = Path.Combine(this.outputDir, Guid.NewGuid() + ".coverage");
+            this.configXmlElement = DynamicCoverageDataCollectorImplTests.CreateXmlElement(ConfigXml)["CodeCoverage"];
+            this.vangurdCommandBuilderMock.Setup(c =>
+                    c.GenerateCommandLine(VangurdCommand.Shutdown, this.sessionName, It.IsAny<string>(), It.IsAny<string>()))
+                .Returns(VangurdTests.GetShutdownCommand(this.sessionName));
+            this.vanguard.Initialize(this.sessionName, this.configFileName, this.configXmlElement, this.dataCollectionLoggerMock.Object);
+            this.collectorUtilityMock.Setup(c => c.GetVanguardPath()).Returns(Path.Combine(Directory.GetCurrentDirectory(), "CodeCoverage.exe"));
+        }
+
+        [TestCleanup]
+        public void Cleanup()
+        {
+            Environment.SetEnvironmentVariable(EnvironmentHelper.VstestConnectionTimeout, "");
+            this.vanguard.Stop();
+            File.Delete(this.configFileName);
+            Directory.Delete(this.outputDir, true);
+        }
+
+        [TestMethod]
+        public void InitializeShouldCreateConfigFile()
+        {
+            Assert.IsTrue(File.Exists(this.configFileName));
+            StringAssert.Contains(
+                VangurdTests.ConfigXml.Replace(" ", string.Empty).Replace(Environment.NewLine, String.Empty),
+                File.ReadAllText(this.configFileName).Replace(" ", string.Empty).Replace(Environment.NewLine, String.Empty));
+        }
+
+        [TestMethod]
+        public void StartShouldStartVanguardProcessWithCollectCommand()
+        {
+            var cts = new CancellationTokenSource();
+            var numOfProcessCreatedTask = NumberOfProcessLaunchedUtility.NumberOfProcessCreated(
+                cts,
+                VangurdTests.CodeCoverageExeFileName);
+
+            this.vangurdCommandBuilderMock.Setup(c =>
+                    c.GenerateCommandLine(VangurdCommand.Collect, It.IsAny<string>(), It.IsAny<string>(), It.IsAny<string>()))
+                .Returns(VangurdTests.GetCollectCommand(this.sessionName, this.outputFileName, this.configFileName));
+
+            this.vanguard.Start(this.outputFileName, this.dataCollectionContext);
+            cts.Cancel();
+
+            // TODO find the reason why more than one processes launched.
```

Comment:
It's possible after CodeCoverage.exe is launched, it is not getting killed, or is in bad state

Reply:
This is happening consistently when running these tests under code coverage. It is not causing either stale process or bad state. However don't exact reason why the addition process coming up for each command.